File: packages/provide-uterm-go/tunnelclient/intercept.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package tunnelclient

import (
	"context"
	"encoding/base64"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// denylistedHeaders are headers an operator-controlled browser MUST NOT inject
// into a forwarded request. They split into three categories — hop-by-hop
// (RFC 7230 §6.1), length/framing (request-smuggling vector), and
// identity/authority (impersonation vector). Mirrors intercept.py's
// _DENYLISTED_HEADERS. Names are lowercased for case-insensitive matching.
var denylistedHeaders = map[string]struct{}{
	// Hop-by-hop
	"connection":          {},
	"keep-alive":          {},
	"proxy-authenticate":  {},
	"proxy-authorization": {},
	"te":                  {},
	"trailer":             {},
	"transfer-encoding":   {},
	"upgrade":             {},
	// Framing / smuggling
	"content-length": {},
	// Identity / authority
	"host":              {},
	"authorization":     {},
	"cookie":            {},
	"forwarded":         {},
	"x-forwarded-for":   {},
	"x-forwarded-host":  {},
	"x-forwarded-proto": {},
	"x-real-ip":         {},
}

// InterceptDecision is the browser's decision for a paused request.
type InterceptDecision struct {
	Action  string            // "forward" | "drop" | "modify"
	Headers map[string]string // non-nil only for a sanitized "modify"
	Body    []byte            // non-nil only when "modify" supplied a new body
}

func defaultDecision(action string) InterceptDecision {
	return InterceptDecision{Action: action}
}

// SanitizeHeaders strips denylisted headers from an operator-supplied header
// map. It returns a new map; the input is not mutated. Dropped names are
// returned (sorted) so the caller can log why a modify didn't take effect
// verbatim. Mirrors intercept.py's _sanitize_headers.
func SanitizeHeaders(raw map[string]string) (cleaned map[string]string, dropped []string) {
	cleaned = make(map[string]string, len(raw))
	for k, v := range raw {
		if _, bad := denylistedHeaders[strings.ToLower(k)]; bad {
			dropped = append(dropped, k)
			continue
		}
		cleaned[k] = v
	}
	sort.Strings(dropped)
	return cleaned, dropped
}

// ParseActionMessage parses an http_action message from the browser into an
// InterceptDecision. Browser-supplied headers are sanitized against the
// denylist. An unknown action falls back to "forward". Invalid base64 bodies
// are ignored (body stays nil). Mirrors intercept.py's parse_action_message.
func ParseActionMessage(msg map[string]any) InterceptDecision {
	action := "forward"
	if a, ok := msg["action"].(string); ok {
		action = a
	}
	if action != "forward" && action != "drop" && action != "modify" {
		action = "forward"
	}
	d := InterceptDecision{Action: action}
	if action != "modify" {
		return d
	}
	if rawHeaders, ok := msg["headers"].(map[string]any); ok {
		src := make(map[string]string, len(rawHeaders))
		for k, v := range rawHeaders {
			src[k] = stringifyHeaderValue(v)
		}
		cleaned, _ := SanitizeHeaders(src)
		d.Headers = cleaned
	}
	if b64, ok := msg["body_b64"].(string); ok {
		if body, err := base64.StdEncoding.DecodeString(b64); err == nil {
			d.Body = body
		}
	}
	return d
}

// stringifyHeaderValue coerces a JSON header value to string like Python's
// str(v). JSON numbers arrive as float64; format an integral value without a
// trailing ".0".
func stringifyHeaderValue(v any) string {
	switch t := v.(type) {
	case string:
		return t
	case float64:
		if t == float64(int64(t)) {
			return strconv.FormatInt(int64(t), 10)
		}
		return strconv.FormatFloat(t, 'g', -1, 64)
	case bool:
		if t {
			return "True"
		}
		return "False"
	default:
		return fmt.Sprintf("%v", t)
	}
}

// InterceptGate manages pending intercepted HTTP requests. It is the Go port of
// intercept.py's InterceptGate, using buffered channels in place of asyncio
// Futures. All fields and the pending map are guarded by mu; AwaitDecision does
// not hold mu while blocking.
type InterceptGate struct {
	mu             sync.Mutex
	enabled        bool
	inspectEnabled bool
	timeoutS       float64
	timeoutAction  string
	pending        map[string]chan InterceptDecision
}

// NewInterceptGate builds a gate. timeoutS is clamped to >= 1.0 and
// timeoutAction is coerced to "forward" unless it is exactly "drop", matching
// the Python constructor.
func NewInterceptGate(timeoutS float64, timeoutAction string) *InterceptGate {
	if timeoutS < 1.0 {
		timeoutS = 1.0
	}
	if timeoutAction != "forward" && timeoutAction != "drop" {
		timeoutAction = "forward"
	}
	return &InterceptGate{
		inspectEnabled: true,
		timeoutS:       timeoutS,
		timeoutAction:  timeoutAction,
		pending:        make(map[string]chan InterceptDecision),
	}
}

// TimeoutS returns the configured per-request wait (seconds).
func (g *InterceptGate) TimeoutS() float64 { return g.timeoutS }

// TimeoutAction returns the configured on-timeout action.
func (g *InterceptGate) TimeoutAction() string { return g.timeoutAction }

// Enabled reports whether interception (pause-before-forward) is on.
func (g *InterceptGate) Enabled() bool {
	g.mu.Lock()
	defer g.mu.Unlock()
	return g.enabled
}

// InspectEnabled reports whether inspection (event emission) is on.
func (g *InterceptGate) InspectEnabled() bool {
	g.mu.Lock()
	defer g.mu.Unlock()
	return g.inspectEnabled
}

// SetEnabled toggles interception.
func (g *InterceptGate) SetEnabled(v bool) {
	g.mu.Lock()
	g.enabled = v
	g.mu.Unlock()
}

// SetInspectEnabled toggles inspection.
func (g *InterceptGate) SetInspectEnabled(v bool) {
	g.mu.Lock()
	g.inspectEnabled = v
	g.mu.Unlock()
}

// PendingCount is the number of requests awaiting a browser decision.
func (g *InterceptGate) PendingCount() int {
	g.mu.Lock()
	defer g.mu.Unlock()
	return len(g.pending)
}

// AwaitDecision blocks until the browser resolves rid, the per-request timeout
// expires, or ctx is cancelled. On timeout/cancel it returns the configured
// timeout-action decision.
func (g *InterceptGate) AwaitDecision(ctx context.Context, rid string) InterceptDecision {
	ch := make(chan InterceptDecision, 1)
	g.mu.Lock()
	g.pending[rid] = ch
	g.mu.Unlock()
	defer func() {
		g.mu.Lock()
		delete(g.pending, rid)
		g.mu.Unlock()
	}()

	timer := time.NewTimer(time.Duration(g.timeoutS * float64(time.Second)))
	defer timer.Stop()
	select {
	case d := <-ch:
		return d
	case <-timer.C:
		return defaultDecision(g.timeoutAction)
	case <-ctx.Done():
		return defaultDecision(g.timeoutAction)
	}
}

// Resolve delivers a browser decision to a pending request. It returns true if
// the request was found (and had not already been resolved).
func (g *InterceptGate) Resolve(rid string, decision InterceptDecision) bool {
	g.mu.Lock()
	ch, ok := g.pending[rid]
	if ok {
		delete(g.pending, rid)
	}
	g.mu.Unlock()
	if !ok {
		return false
	}
	ch <- decision // buffered (cap 1) — never blocks
	return true
}

// CancelAll resolves every pending request with action and returns the count.
func (g *InterceptGate) CancelAll(action string) int {
	g.mu.Lock()
	defer g.mu.Unlock()
	n := 0
	for rid, ch := range g.pending {
		ch <- defaultDecision(action)
		delete(g.pending, rid)
		n++
	}
	return n
}

File: packages/provide-uterm-server/src/provide/uterm/tunnel/intercept.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""InterceptGate — pause/resume state machine for HTTP request interception."""

from __future__ import annotations

import asyncio
import base64
import logging
from typing import Any, TypedDict

logger = logging.getLogger(__name__)

# Headers that an operator-controlled browser MUST NOT be allowed to inject
# into the forwarded request. These split into three categories:
#
#   - Hop-by-hop (RFC 7230 §6.1): connection-scoped, must not be proxied.
#   - Length / framing: reflecting client-controlled values enables HTTP
#     request smuggling against downstream servers. httpx2 computes
#     Content-Length from the body anyway.
#   - Identity / authority: forwarding these lets the operator impersonate
#     the original requester or hijack downstream authentication, which is
#     the threat the interceptor itself was meant to mitigate.
#
# Names are lowercased for case-insensitive matching.
_DENYLISTED_HEADERS: frozenset[str] = frozenset(
    {
        # Hop-by-hop
        "connection",
        "keep-alive",
        "proxy-authenticate",
        "proxy-authorization",
        "te",
        "trailer",
        "transfer-encoding",
        "upgrade",
        # Framing / smuggling
        "content-length",
        # Identity / authority
        "host",
        "authorization",
        "cookie",
        "forwarded",
        "x-forwarded-for",
        "x-forwarded-host",
        "x-forwarded-proto",
        "x-real-ip",
    }
)


def _sanitize_headers(raw: dict[str, str]) -> dict[str, str]:
    """Strip denylisted headers from an operator-supplied header dict.

    Returns a new dict; the original is not mutated. Logs the names of any
    headers that were dropped so the operator can see why their modify
    decision didn't take effect verbatim.
    """
    cleaned: dict[str, str] = {}
    dropped: list[str] = []
    for key, value in raw.items():
        if key.lower() in _DENYLISTED_HEADERS:
            dropped.append(key)
            continue
        cleaned[key] = value
    if dropped:
        logger.warning("intercept_headers_denylisted dropped=%s", sorted(dropped))
    return cleaned


class InterceptDecision(TypedDict):
    """Browser's decision for a paused request."""

    action: str  # "forward" | "drop" | "modify"
    headers: dict[str, str] | None
    body: bytes | None


def _default_decision(action: str) -> InterceptDecision:
    return InterceptDecision(action=action, headers=None, body=None)


def parse_action_message(msg: dict[str, Any]) -> InterceptDecision:
    """Parse an http_action message from the browser into an InterceptDecision.

    Browser-supplied headers are sanitized against ``_DENYLISTED_HEADERS``
    before reaching the upstream proxy. See the module-level comment for
    why each entry is dropped.
    """
    action = str(msg.get("action", "forward"))
    if action not in ("forward", "drop", "modify"):
        action = "forward"
    headers: dict[str, str] | None = None
    body: bytes | None = None
    if action == "modify":
        raw_headers = msg.get("headers")
        if isinstance(raw_headers, dict):
            headers = _sanitize_headers({str(k): str(v) for k, v in raw_headers.items()})
        body_b64 = msg.get("body_b64")
        if isinstance(body_b64, str):
            try:
                body = base64.b64decode(body_b64, validate=True)
            except Exception:
                logger.warning("intercept_invalid_body_b64 id=%s", msg.get("id"))
    return InterceptDecision(action=action, headers=headers, body=body)


class InterceptGate:
    """Manages pending intercepted HTTP requests as asyncio Futures."""

    def __init__(self, timeout_s: float = 30.0, timeout_action: str = "forward") -> None:
        self.enabled: bool = False
        self.inspect_enabled: bool = True
        self.timeout_s: float = max(1.0, timeout_s)
        self.timeout_action: str = timeout_action if timeout_action in ("forward", "drop") else "forward"
        self._pending: dict[str, asyncio.Future[InterceptDecision]] = {}

    @property
    def pending_count(self) -> int:
        """Number of requests currently awaiting a browser decision."""
        return len(self._pending)

    async def await_decision(self, rid: str) -> InterceptDecision:
        """Block until browser sends http_action or timeout expires."""
        loop = asyncio.get_running_loop()
        fut: asyncio.Future[InterceptDecision] = loop.create_future()
        self._pending[rid] = fut
        try:
            return await asyncio.wait_for(fut, timeout=self.timeout_s)
        except TimeoutError:
            return _default_decision(self.timeout_action)
        finally:
            self._pending.pop(rid, None)

    def resolve(self, rid: str, decision: InterceptDecision) -> bool:
        """Resolve a pending request. Returns True if found and resolved."""
        fut = self._pending.get(rid)
        if fut is None or fut.done():
            return False
        fut.set_result(decision)
        return True

    def cancel_all(self, action: str = "forward") -> int:
        """Resolve all pending requests. Returns count resolved."""
        decision = _default_decision(action)
        count = 0
        for fut in self._pending.values():
            if not fut.done():
                fut.set_result(decision)
                count += 1
        self._pending.clear()
        return count
